### ISSUE TITLE
Brace pairing check in ec_glob.c incorrect?

### ISSUE THREAD

**Issue by craigbarnes** [editorconfig/editorconfig-core-c#50]
I was just reading the code in `ec_glob.c` and noticed [this loop] doesn't seem to be doing what the variable name would suggest. Checking for equal numbers of opening and closing braces doesn't necessarily mean they're "paired". It's not immediately obvious if or how this effects the correctness of the code, but I just thought I'd point it out in case it's a bug.

[this loop]: https://github.com/editorconfig/editorconfig-core-c/blob/a8fae2375e67887bd73baa11d921339f24e8b9a9/src/lib/ec_glob.c#L89-L107

**Comment by craigbarnes** [comment]
It seems if the braces are (incorrectly) determined to be "paired", all braces in the pattern get [escaped] and only represent themselves. That seems to prevent e.g. `}*.{x,y}` from matching `}file.y`. It also seems like it could cause false matches (and [signed integer overflow]?) as well in the case of something like `}foo{a,b`, which has even (but not paired) braces.

[escaped]: https://github.com/editorconfig/editorconfig-core-c/blob/a8fae2375e67887bd73baa11d921339f24e8b9a9/src/lib/ec_glob.c#L218-L222

[signed integer overflow]: https://github.com/editorconfig/editorconfig-core-c/blob/a8fae2375e67887bd73baa11d921339f24e8b9a9/src/lib/ec_glob.c#L294

**Comment by xuhdev** [comment]
Thanks for your many bug reports! How does #54 look to you?

**Comment by craigbarnes** [comment]
It seems like #54 should fix the signed integer overflow, but it still looks to make the escaping more aggressive that necessary. In most regex implementations, an unbalanced closing parenthesis is not considered a special character at all and just represents itself.

The approach I took is to [find](https://github.com/craigbarnes/dte/blob/66984b996e48ae79ced3f9b80d40190602fd02f1/src/editorconfig/match.c#L27-L49) the last index at which there is a paired, closing brace and beyond that point just [escape](https://github.com/craigbarnes/dte/blob/66984b996e48ae79ced3f9b80d40190602fd02f1/src/editorconfig/match.c#L130-L133) all braces. Closing braces _before_ that point also get escaped if `brace_level == 0` (i.e. there's no opening brace to close -- same as in regex).

**Comment by craigbarnes** [comment]
By the way, scratch what I said about signed integer overflow. The incorrect pairing check doesn't actually lead to overflow -- it just allows `brace_level` to possibly go negative, which seems like it would be harmless. I guess the real issue was just false matches, which #54 seems to also solve. The point about over-escaping still stands though.

**Comment by craigbarnes** [comment]
I've been thinking about this some more and I'm starting to think your approach of "over-escaping" might actually be the best thing to do. Either that or just make patterns with invalid brace pairing match nothing (always return `false`). 

I just noticed that my brace pairing check allows through _unclosed_ braces if they are nested but the inner level is closed, e.g. `{a,b{c,d{e}`. It's not really obvious what the semantics should be in this case. The regex equivalent just causes `regcomp()` to return a `REG_EPAREN` error code. I guess if people supply badly formed patterns they can't really expect it to match anything.

tl;dr I think your PR #54 seems fine. My only remaining question is should badly paired braces just cause the pattern to match nothing at all instead?

**Comment by xuhdev** [comment]
I'm personally leaning toward not making incompatible changes unless there is a compelling reason. In this case, it would be keep unpaired curly braces match. What do you think?

**Comment by craigbarnes** [comment]
I think edge cases like this are more a question of ensuring they don't lead to undefined behaviour or exploits than the actual matching semantics. It seems unlikely that any real `.editorconfig` files are relying on this edge case. That said, it doesn't really seem to matter one way or the other -- if people want to match literal braces all they need do is escape them. I'll just copy whatever behaviour upstream decides on.

**Comment by xuhdev** [comment]
Fixed by #54
